RFC-0009/RFC-0009-binary-exchange.md

```diff
@@ -0,0 +1,265 @@
+stopping and that associated resources can be freed. Delete requests are acknowledged.
+
+Requests and replies are implemented as binary encoded protocol data units on top of TCP.
+This minimizes the protocol overhead.
+
+### Configuration Options
+
+BinX has two configuration properties that are part of the Prestissimo (native) worker
+configuration:
+
+|Property name | Type| Values| Effect |
+|--------|------|------|------|
+|binx.enabled|bool| True or false, defaults to false | Enables BinX when set to true|
+|binx.server.port|uint64_t|Valid port number > 1000, defaults to 8091| The port number used by BinX|
+
+The configuration must be homogeneous across all worker nodes. BinX is disabled by default
+and must be explicitly enabled.
+
+If enabled, the worker's logs will contain the following messages:
+```
+I20240828 14:57:11.361124    19 PrestoServer.cpp:599] [PRESTO_STARTUP] Binary Server IO executor 'BinSrvIO' has 64 threads.
+I20240828 14:57:11.364262    19 BinaryExchangeServer.cpp:203] Starting binary exchange server on port 8091
+```
+
+## Implementation Overview
+
+The implementation covers the protocol design, the BinX server implementation and the BinX exchange
+source implementation.
+
+### Binary Exchange Protocol
+
+The protocol consists of requests and replies that are implemented by the `BinRequest` and the `BinReply` classes defined
+in `BinRequestReply.h`. All three request types for data, acknowledge and delete share the same structure.
+A request consists of the following fields:
+|Field name| Type | Semantic |
+| ---- | ---- | ---- |
+|requestType| enum (DATA, ACK, DELETE) | The type of request |
+|getDataSize| bool | When true, requests the sizes of the remaining available data pages; when false requests data |
+|maxSizeOctets| uint64_t | The maximum allowed size of the data pages|
+|maxWaitMicroSeconds|uint64_t| The maximum wait time in microseconds|
+|taskId| std::string | The unique ID of the remote task from which data is requested|
+|bufferId| uint64_t | The ID of the buffer within the remote task|
+|pageToken| uint64_t | The sequence number of the data page that is requested|
+
+Acknowledge and delete requests don't need all fields and set `getDataSize`, `maxSizeOctets`,
+and `maxWaitMicroSeconds` to false and 0, respectively.
+
+A reply has the following fields:
+
+|Field name| Type | Semantic |
+| ---- | ---- | ---- |
+|replyType| enum (DATA, ACK, DELETE) | The type of reply |
+|status| enum (OK, NODATA, SERVER_ERROR, TIMEOUT) | The reply status |
+|bufferComplete| bool | True if the entire buffer has been transferred |
+|pageToken|uint64_t| The token (sequence number) of the first data page in the reply|
+|pageNextToken|uint64_t| The token (sequence number) of the next available page that can be requested|
+|remainingBytes| std::vector<uint64_t> | Contains the sizes of available pages when requested|
+|data| IOBuf | Contains the data payload |
```

Comment:
Slightly ignorant question, but by encoding this field parallel to the other fields, does it require a memcpy?